### PR TITLE
169532131 add script check circ deps

### PR DESCRIPTION
**Pull request by giskarda**

### CODE REVIEW

File: utils/README.md [hunk 1]
```diff
@@ -80,3 +80,20 @@ To synchronize users run:
...
+## Check for circular dependencies in live modules
+
+This is useful if we are building an environment from scratch after a long while.
+
+### Usage
+
+1. Create a python virtual environment
+2. Enable the environment
+2. Install required libries
+3. Run the script
```
Comment (LucaPrete) [inline utils/README.md]: Please, use * for lists instead to be consistent with the  format of the rest of the READMEs

File: utils/README.md [hunk 2]
```diff
@@ -80,3 +80,20 @@ To synchronize users run:
 ```shell
 source az-apim-sync.sh
 ```
+
+## Check for circular dependencies in live modules
+
+This is useful if we are building an environment from scratch after a long while.
```
Comment (LucaPrete) [inline utils/README.md]: This will be used primarily in the CI process. Can we put instead what the script does?

File: utils/README.md [hunk 3]
```diff
@@ -80,3 +80,20 @@ To synchronize users run:
...
+
+### Usage
+
+1. Create a python virtual environment
+2. Enable the environment
+2. Install required libries
+3. Run the script
+
+### Example:
```
Comment (LucaPrete) [inline utils/README.md]: Can you please substitute this with a unique block as the one below?
```
...
```

File: utils/check_live_circular_dependencies.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+import hcl
+
+dot = Digraph(comment='tfvars dev')
+
+parser = argparse.ArgumentParser(description='Scan live environment repository for circular dependencies')
+parser.add_argument('-d', '--directory', help='relative path')
+parser.add_argument('-o', '--output', help='output file')
+
+args = parser.parse_args()
+
```
Comment (LucaPrete) [inline utils/check_live_circular_dependencies.py]: please remove one of the two empty lines

File: utils/check_live_circular_dependencies.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+            with open(os.path.join(root, name), 'r') as fp:
+                obj = hcl.load(fp)
+                try:
+                    dot.node(os.path.join(root,name), os.path.join(root,name))
+                    dependencies = obj["terragrunt"]["dependencies"]["paths"]
+                    for dep in dependencies:
+                        dot.edge(os.path.join(root, name), dep)
+                except KeyError:
+                    dot.node(name, name)
+
```
Comment (LucaPrete) [inline utils/check_live_circular_dependencies.py]: please remove one of the two empty lines

File: utils/check_live_circular_dependencies.py (hunk 4 above)
Comment (ioggstream) [reply utils/check_live_circular_dependencies.py]: @LucaPrete @giskarda can't we delegate all that syntax stuff  to `black` ?

Comment (giskarda) [reply utils/check_live_circular_dependencies.py]: sure thing.

**Comment by giskarda** [comment]
I'm going to answer in line here (I dont think I can comment your comment)

> this change doesn't come from the latest master! Please, use the updated master as starting point of your change

Ok!

> the script parses also the .terragrunt-cache folder. This must be avoided.

Fixed.

> The script produces a graph that is barely readable - either we fix the graph or the feature is useless

What you dont like? Suggestions? 

> vnet_common should depend by resource group, but I don't see any arrow

../dev/westeurope/vnet_common/terraform.tfvars has a straight line to resource_group. I can confirm

> The most important feature of the script should be find if any circular dependency is found, returning some error and exiting with -1. This will be specifically useful for CI pipelines, executed when new TF files are written. IMHO this is the primary feature.i

I've added an as_text() function that returns 1 when an circular dependency or an issue is found.

>The graph instead is a very nice to have feature, that should become optional. 

Ok 

> you should also add the following exclusions in the gitignore:
utils/live
utils/live.pdf
utils/venv/

This depends on what the user wants as a name and how it likes.

**Comment by LucaPrete** [comment]
Answers **inline**

> this change doesn't come from the latest master! Please, use the updated master as starting point of your change

Ok!

> the script parses also the .terragrunt-cache folder. This must be avoided.

Fixed.

> The script produces a graph that is barely readable - either we fix the graph or the feature is useless

What you dont like? Suggestions? 
**I'll send you a screnshot of what I see on my screen to let you better understand what I mean. The graph is barely visible and you have to scroll horizzontally back and forth to see associations between the vertex of the graph. Maybe removing .terragrunt-cached folders made things a little bit better**

> vnet_common should depend by resource group, but I don't see any arrow

../dev/westeurope/vnet_common/terraform.tfvars has a straight line to resource_group. I can confirm

**ok**

> The most important feature of the script should be find if any circular dependency is found, returning some error and exiting with -1. This will be specifically useful for CI pipelines, executed when new TF files are written. IMHO this is the primary feature.

I've added an as_text() function that returns 1 when an circular dependency or an issue is found.

**Does it also say what the circular dependencies are?**

>The graph instead is a very nice to have feature, that should become optional. 

Ok 

> you should also add the following exclusions in the gitignore:
utils/live
utils/live.pdf
utils/venv/

This depends on what the user wants as a name and how it likes.

**venv would be a constant I guess. let's put instead of the live files in a chart (any better idea?). Then, let's add that and venv to .gitignore **

**Pull request merged**